includes/class-meta-pixel.php

```diff
@@ -0,0 +1,228 @@
+		$sanitized = [];
+		if ( ! is_array( $value ) ) {
+			return $defaults;
+		}
+		if ( isset( $value['active'] ) ) {
+			$sanitized['active'] = (bool) $value['active'];
+		} else {
+			$sanitized['active'] = $defaults['active'];
+		}
+		if ( isset( $value['pixel_id'] ) ) {
+			$sanitized['pixel_id'] = (string) $value['pixel_id']; // store as string to avoid PHP_INT_MAX problems.
+		} else {
+			$sanitized['pixel_id'] = $defaults['pixel_id'];
+		}
+		return $sanitized;
+	}
+
+	/**
+	 * Checks if AMP plugin is enabled and if the current response is being server as AMP
+	 *
+	 * @return boolean
+	 */
+	public static function is_amp() {
+		return function_exists( 'is_amp_endpoint' ) && is_amp_endpoint();
+	}
+
+	/**
+	 * Gets the template for the img tag snippet
+	 *
+	 * @return string
+	 */
+	public static function get_img_snippet() {
+		global $wp;
+		$current_user = wp_get_current_user();
+		$event_params = [
+			'page_title' => get_the_title(),
+			'user_role'  => empty( $current_user->roles ) ? 'guest' : $current_user->roles[0],
+			'event_url'  => home_url( $wp->request ),
+		];
+
+		if ( is_single() ) {
+			$event_params['post_type'] = get_post_type();
+			$event_params['post_id']   = get_the_ID();
+		}
+
+		$url_params = http_build_query( [ 'cd' => $event_params ] );
+
+		return '<img height="1" width="1" style="display: none;" src="https://www.facebook.com/tr?id=__PIXEL_ID__&ev=PageView&noscript=1&' . $url_params . '">';
+	}
+
+	/**
+	 * Gets the template for the script tag snippet
+	 *
+	 * @return string
+	 */
+	public static function get_script_snippet() {
+		return "
+		<script>
+		!function(f,b,e,v,n,t,s)
+		{if(f.fbq)return;n=f.fbq=function(){n.callMethod?
+		n.callMethod.apply(n,arguments):n.queue.push(arguments)};
+		if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
+		n.queue=[];t=b.createElement(e);t.async=!0;
+		t.src=v;s=b.getElementsByTagName(e)[0];
+		s.parentNode.insertBefore(t,s)}(window, document,'script',
+		'https://connect.facebook.net/en_US/fbevents.js');
+		fbq('init', '__PIXEL_ID__');
+		fbq('track', 'PageView');
+		</script>
+		";
+	}
+
+	/**
+	 * Prints sinppets in the header
+	 *
+	 * @return void
+	 */
+	public static function print_head_snippet() {
+		if ( self::is_amp() ) {
+			return;
+		}
+		$pixel_id = self::get_pixel_id();
+		if ( empty( $pixel_id ) ) {
+			return;
+		}
+		echo str_replace( '__PIXEL_ID__', intval( $pixel_id ), self::get_script_snippet() ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+	}
+
+	/**
+	 * Prints snippets in the footer
+	 *
+	 * @return void
+	 */
+	public static function print_footer_snippet() {
+		$pixel_id = self::get_pixel_id();
+		if ( empty( $pixel_id ) ) {
+			return;
+		}
+		// If AMP plugin is enabled, it will convert the image into a <amp-pixel> tag.
+		echo '<noscript>' . str_replace( '__PIXEL_ID__', intval( $pixel_id ), self::get_img_snippet() ) . '</noscript>'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```

Comment:
AMP enabled does not mean the browser doesn't support javascript. A `! self::is_amp()` check is more suitable for this.

Reply:
This is not a check for AMP enabled.

The `noscript` tag is there for non-AMP site. If visiting a non AMP site without JS, the IMG tag will be used for tracking.

If AMP is enabled, the AMP plugin will capture this and replace the IMG with a `<amp-pixel>` tag. So this covers both cases and we don't need to manually add the amp tag. We leave it to the AMP plugin